File: test_security_issues.js
/**
 * Test file with intentional security vulnerabilities for SonarCloud SARIF testing.
 * This file contains various JavaScript security issues that SonarCloud should detect.
 */

// Security Issue 1: Hardcoded credentials
const API_KEY = "sk-1234567890abcdef"; // SonarCloud should flag this
const DATABASE_URL = "mongodb://admin:password123@localhost:27017/mydb";
const JWT_SECRET = "my_super_secret_key";

// Security Issue 2: eval() usage - code injection vulnerability
function executeUserCode(userInput) {
    return eval(userInput); // Dangerous - arbitrary code execution
}

// Security Issue 3: Insecure random number generation
function generateSessionId() {
    return Math.random().toString(36); // Not cryptographically secure
}

// Security Issue 4: SQL injection vulnerability
function getUserById(id) {
    const query = `SELECT * FROM users WHERE id = ${id}`; // SQL injection risk
    return query;
}

// Security Issue 5: XSS vulnerability - innerHTML with user data
function displayUserData(userData) {
    document.getElementById("user-info").innerHTML = userData; // XSS risk
}

// Security Issue 6: Insecure cookie settings
function setAuthCookie(token) {
    document.cookie = `auth_token=${token}; path=/`; // Missing secure, httpOnly, sameSite
}

// Security Issue 7: Weak encryption/hashing
const crypto = require('crypto');
function hashPassword(password) {
    return crypto.createHash('md5').update(password).digest('hex'); // MD5 is weak
}

// Security Issue 8: Path traversal vulnerability
const fs = require('fs');
function readUserFile(filename) {
    return fs.readFileSync(`/var/uploads/${filename}`, 'utf8'); // No path validation
}

// Security Issue 9: Command injection
const { exec } = require('child_process');
function processFile(filename) {
    exec(`convert ${filename} output.jpg`); // Command injection risk
}

// Security Issue 10: Insecure deserialization
function loadUserSession(sessionData) {
    return JSON.parse(sessionData); // Should validate/sanitize first
}

// Security Issue 11: Information disclosure in error messages
function authenticateUser(username, password) {
    if (!userExists(username)) {
        throw new Error(`User ${username} does not exist`); // Username disclosure
    }
    if (!isValidPassword(username, password)) {
        throw new Error(`Invalid password for user ${username}`); // Username disclosure
    }
}

// Security Issue 12: Insecure HTTP requests
const https = require('https');
function makeApiCall(url) {
    const options = {
        rejectUnauthorized: false // Disables certificate validation
    };
    return https.get(url, options);
}

// Security Issue 13: Prototype pollution vulnerability
function mergeObjects(target, source) {
    for (let key in source) {
        target[key] = source[key]; // No protection against __proto__ pollution
    }
    return target;
}

// Security Issue 14: Missing input validation
function transferFunds(fromAccount, toAccount, amount) {
    // No validation on amount, could be negative
    return processTransfer(fromAccount, toAccount, amount);
}

// Security Issue 15: Timing attack vulnerability
function comparePasswords(input, stored) {
    if (input.length !== stored.length) {
        return false; // Early return reveals length information
    }
    for (let i = 0; i < input.length; i++) {
        if (input[i] !== stored[i]) {
            return false; // Early return enables timing attacks
        }
    }
    return true;
}

// Security Issue 16: Debug information exposure
const DEBUG = true;
if (DEBUG) {
    console.log(`Database URL: ${DATABASE_URL}`); // Credential exposure
    console.log(`JWT Secret: ${JWT_SECRET}`);
}

// Dummy functions to make the code syntactically correct
function userExists(username) { return true; }
function isValidPassword(username, password) { return false; }
function processTransfer(from, to, amount) { return "processed"; }

console.log("This file contains intentional security vulnerabilities for testing");
console.log("Do not use this code in production!");

File: test_security_issues.py
#!/usr/bin/env python3
"""
Test file with intentional security vulnerabilities for SonarCloud SARIF testing.
This file contains various security issues that SonarCloud should detect.
"""

import os
import subprocess
import hashlib
import random
import pickle

# Security Issue 1: Hardcoded credentials
DATABASE_PASSWORD = "admin123"  # SonarCloud should flag this
API_KEY = "sk-1234567890abcdef"
SECRET_TOKEN = "supersecret"

# Security Issue 2: SQL Injection vulnerability
def get_user_data(user_id):
    query = f"SELECT * FROM users WHERE id = {user_id}"  # SQL injection risk
    # This would be vulnerable in a real database query
    return query

# Security Issue 3: Command injection vulnerability  
def run_command(user_input):
    # Direct command execution with user input
    result = subprocess.run(f"ls {user_input}", shell=True)  # Command injection
    return result

# Security Issue 4: Weak cryptographic hash
def hash_password(password):
    return hashlib.md5(password.encode()).hexdigest()  # MD5 is weak

# Security Issue 5: Insecure random number generation
def generate_token():
    return str(random.random())  # Not cryptographically secure

# Security Issue 6: Pickle deserialization vulnerability
def load_user_data(data):
    return pickle.loads(data)  # Unsafe deserialization

# Security Issue 7: Path traversal vulnerability
def read_file(filename):
    with open(f"/var/data/{filename}", 'r') as f:  # No path validation
        return f.read()

# Security Issue 8: Weak SSL/TLS configuration
import ssl
def create_context():
    context = ssl.create_default_context()
    context.check_hostname = False  # Disables hostname verification
    context.verify_mode = ssl.CERT_NONE  # Disables certificate verification
    return context

# Security Issue 9: Debug information disclosure
DEBUG = True
if DEBUG:
    print(f"Database password: {DATABASE_PASSWORD}")  # Information disclosure

# Security Issue 10: Insecure temporary file creation
import tempfile
def create_temp_file():
    # Creates file with predictable name and insecure permissions
    temp_path = "/tmp/myapp_" + str(random.randint(1000, 9999))
    with open(temp_path, 'w') as f:
        f.write("sensitive data")
    os.chmod(temp_path, 0o777)  # World writable
    return temp_path

# Security Issue 11: LDAP injection
def authenticate_user(username):
    ldap_query = f"(&(uid={username})(objectClass=person))"  # LDAP injection
    return ldap_query

# Security Issue 12: XXE vulnerability simulation
def parse_xml(xml_data):
    import xml.etree.ElementTree as ET
    # Parsing XML without disabling external entities
    root = ET.fromstring(xml_data)  # Vulnerable to XXE
    return root

if __name__ == "__main__":
    print("This file contains intentional security vulnerabilities for testing")
    print("Do not use this code in production!")